"try expression alternatives have incompatible types" is confusing wording

Simple example:
```rust
fn foo(x: Result<i32, ()>) -> Result<(), ()> {
    let y: u32 = x?;
    Ok(())
}
```
Compilation result (`rust version 1.44.0-nightly (52fa23add 2020-04-18)`):
```
error[E0308]: try expression alternatives have incompatible types
 --> src/lib.rs:2:18
  |
2 |     let y: u32 = x?;
  |                  ^^ expected `u32`, found `i32`
  |
help: you can convert an `i32` to `u32` and panic if the converted value wouldn't fit
  |
2 |     let y: u32 = x?.try_into().unwrap();
  |                  ^^^^^^^^^^^^^^^^^^^^^^
```

Parts of this diagnostic are great, showing the mismatched types (`u32` and `i32`) and even suggesting a valid possible fix.  But the wording of the error itself is confusing.  What "try expression alternatives"?  I don't see any alternatives.

Taking a glance at the compiler implementation, it's referring to the arms of the `match` block that `?` desugars into.  But it's confusing to talk about those, because:
- They're not visible to the user.
- Why would there be a type mismatch between the arms in the first place, if there are only two arms and one of them always ends in a `return` rather than producing a value?  Did type inference infer the error branch as having type `u32` due to the context?

  If so, the message is technically accurate, but unhelpful: from the user's point of view, the 'real' issue is a mismatch between the expected type of `x?` (namely `i32`) and the context.  As far as I can imagine, this is the *only* possible cause of a type mismatch between the arms of a desugared `?`.  So it seems like it would be fine to just change the error message to not mention the arms.

As a sidenote, if I take the `-Zunpretty=hir` output and try to compile that:
```rust
#![feature(try_trait)]
use std::ops::Try;
use std::convert::From;
fn foo(x: Result<i32, ()>) -> Result<(), ()> {
    let y: u32 = match Try::into_result(x) {
        Ok(val) => val,
        Err(err) => return Try::from_error(From::from(err)),
    };
    Ok(())
}
```

...oddly enough, I don't get a "match arms have incompatible types" error, but a more helpful one instead:
```
error[E0308]: mismatched types
 --> q.rs:6:20
  |
6 |         Ok(val) => val,
  |                    ^^^ expected `u32`, found `i32`
  |
help: you can convert an `i32` to `u32` and panic if the converted value wouldn't fit
  |
6 |         Ok(val) => val.try_into().unwrap(),
  |                    ^^^^
```

--- a/compiler/rustc_errors/src/diagnostic.rs
+++ b/compiler/rustc_errors/src/diagnostic.rs
@@ -74,12 +74,24 @@ impl DiagnosticStyledString {
     pub fn highlighted<S: Into<String>>(t: S) -> DiagnosticStyledString {
         DiagnosticStyledString(vec![StringPart::Highlighted(t.into())])
     }
+
+    pub fn content(&self) -> String {
+        self.0.iter().map(|x| x.content()).collect::<String>()
+    }
 }
 
 #[derive(Debug, PartialEq, Eq)]
 pub enum StringPart {
     Normal(String),
     Highlighted(String),
+}
+
+impl StringPart {
+    pub fn content(&self) -> &str {
+        match self {
+            &StringPart::Normal(ref s) | &StringPart::Highlighted(ref s) => s,
+        }
+    }
 }
 
 impl Diagnostic {

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1971,6 +1971,8 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         trace: TypeTrace<'tcx>,
         terr: &TypeError<'tcx>,
     ) -> DiagnosticBuilder<'tcx> {
+        use crate::traits::ObligationCauseCode::MatchExpressionArm;
+
         debug!("report_and_explain_type_error(trace={:?}, terr={:?})", trace, terr);
 
         let span = trace.cause.span(self.tcx);
@@ -2011,6 +2013,19 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                             }
                         }
                         _ => {}
+                    }
+                }
+                if let MatchExpressionArm(box MatchExpressionArmCause { source, .. }) =
+                    trace.cause.code
+                {
+                    if let hir::MatchSource::TryDesugar = source {
+                        if let Some((expected_ty, found_ty)) = self.values_str(trace.values) {
+                            err.note(&format!(
+                                "`?` operator cannot convert from `{}` to `{}`",
+                                found_ty.content(),
+                                expected_ty.content(),
+                            ));
+                        }
                     }
                 }
                 err
@@ -2585,9 +2600,7 @@ impl<'tcx> ObligationCauseExt<'tcx> for ObligationCause<'tcx> {
             CompareImplTypeObligation { .. } => Error0308("type not compatible with trait"),
             MatchExpressionArm(box MatchExpressionArmCause { source, .. }) => {
                 Error0308(match source {
-                    hir::MatchSource::TryDesugar => {
-                        "try expression alternatives have incompatible types"
-                    }
+                    hir::MatchSource::TryDesugar => "`?` operator has incompatible types",
                     _ => "`match` arms have incompatible types",
                 })
             }

--- a/src/test/ui/inference/issue-71309.rs
+++ b/src/test/ui/inference/issue-71309.rs
@@ -1,0 +1,7 @@
+fn foo(x: Result<i32, ()>) -> Result<(), ()> {
+    let y: u32 = x?;
+    //~^ ERROR: `?` operator has incompatible types
+    Ok(())
+}
+
+fn main() {}

--- a/src/test/ui/inference/issue-71309.stderr
+++ b/src/test/ui/inference/issue-71309.stderr
@@ -1,0 +1,15 @@
+error[E0308]: `?` operator has incompatible types
+  --> $DIR/issue-71309.rs:2:18
+   |
+LL |     let y: u32 = x?;
+   |                  ^^ expected `u32`, found `i32`
+   |
+   = note: `?` operator cannot convert from `i32` to `u32`
+help: you can convert an `i32` to a `u32` and panic if the converted value doesn't fit
+   |
+LL |     let y: u32 = x?.try_into().unwrap();
+   |                    ++++++++++++++++++++
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/issues/issue-51632-try-desugar-incompatible-types.rs
+++ b/src/test/ui/issues/issue-51632-try-desugar-incompatible-types.rs
@@ -6,7 +6,7 @@ fn missing_discourses() -> Result<isize, ()> {
 
 fn forbidden_narratives() -> Result<isize, ()> {
     missing_discourses()?
-    //~^ ERROR try expression alternatives have incompatible types
+    //~^ ERROR: `?` operator has incompatible types
 }
 
 fn main() {}

--- a/src/test/ui/issues/issue-51632-try-desugar-incompatible-types.stderr
+++ b/src/test/ui/issues/issue-51632-try-desugar-incompatible-types.stderr
@@ -1,9 +1,10 @@
-error[E0308]: try expression alternatives have incompatible types
+error[E0308]: `?` operator has incompatible types
   --> $DIR/issue-51632-try-desugar-incompatible-types.rs:8:5
    |
 LL |     missing_discourses()?
    |     ^^^^^^^^^^^^^^^^^^^^^ expected enum `Result`, found `isize`
    |
+   = note: `?` operator cannot convert from `isize` to `Result<isize, ()>`
    = note: expected enum `Result<isize, ()>`
               found type `isize`
 help: try removing this `?`

